activesupport/lib/active_support/messages/rotator.rb

```diff
@@ -6,44 +6,44 @@ module Rotator # :nodoc:
       def initialize(*secrets, on_rotation: nil, **options)
         super(*secrets, **options)
 
-        @options   = options
+        @secrets = secrets
+        @options = options
         @rotations = []
         @on_rotation = on_rotation
       end
 
       def rotate(*secrets, **options)
-        @rotations << build_rotation(*secrets, @options.merge(options))
+        fall_back_to build_rotation(*secrets, **options)
       end
 
-      module Encryptor
+      def fall_back_to(fallback)
+        @rotations << fallback
+        self
+      end
+
+      module Encryptor # :nodoc:
         include Rotator
 
         def decrypt_and_verify(*args, on_rotation: @on_rotation, **options)
           super
         rescue MessageEncryptor::InvalidMessage, MessageVerifier::InvalidSignature
           run_rotations(on_rotation) { |encryptor| encryptor.decrypt_and_verify(*args, **options) } || raise
         end
-
-        private
-          def build_rotation(secret = @secret, sign_secret = @sign_secret, options)
-            self.class.new(secret, sign_secret, **options)
-          end
       end
 
-      module Verifier
+      module Verifier # :nodoc:
         include Rotator
 
         def verified(*args, on_rotation: @on_rotation, **options)
           super || run_rotations(on_rotation) { |verifier| verifier.verified(*args, **options) }
         end
-
-        private
-          def build_rotation(secret = @secret, options)
-            self.class.new(secret, **options)
-          end
       end
 
       private
+        def build_rotation(*secrets, **options)
+          self.class.new(*secrets, *@secrets.drop(secrets.length), **@options, **options)
```

Comment:
I could not get this `drop` here. Why is it necessary?

Reply:
For `MessageEncryptor`, both `secrets` and `@secrets` can be either [a single secret for both encryption and signing, or separate secrets for encryption and signing](https://github.com/rails/rails/blob/83f32f62fe353f12c05ce89029c85db48df264f0/activesupport/lib/active_support/message_encryptor.rb#L148).  Meaning there are four cases we need to handle:

1. single `secrets` + single `@secrets`
2. single `secrets` + separate `@secrets`
3. separate `secrets` + single `@secrets`
4. separate `secrets` + separate `@secrets`

`@secrets.drop(secrets.length)` handles all four cases in the same way that [`Rotator::Encryptor#build_rotation` did](https://github.com/rails/rails/pull/44179/files#diff-9c293303dbc61fa3b1adf390cf32d0e173e8e18333fcf06831e1cba70e1f5ec3L28-L30).  The difference is that `Rotator::Encryptor#build_rotation` used `@secret` and `@sign_secret` instance variables which are only defined [in `MessageEncryptor` itself](https://github.com/rails/rails/blob/83f32f62fe353f12c05ce89029c85db48df264f0/activesupport/lib/active_support/message_encryptor.rb#L149-L150).  That seemed like an encapsulation violation to me, so I switched to `@secrets` which *is* defined in the `Rotator` module.